### PR TITLE
Improve muon analysis plotting

### LINKED ISSUE #26884: Improve muon plotting
The plotting in the muon GUI's should be updated so it can use tile plots

This should be added to both frequency and muon analysis.

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR improves the plotting within the Muon analysis GUI. The primary improvement is the addition of tiled plotting, which enables the plotting of selected detector groups in the same plot window. The plots can be tiled by group/pair or by run (if multiple runs are loaded). Also added is the ability to overlay pairs, or groups, onto the same plot.

Excuse the messy git history, the plotting (and in general the GUI) was changed frequently based on feedback received from the scientists throughout the development of these improvements. 

**To test**
```
Example workflow
1. Open the Muon Analysis 2 GUI
2. Load in some data (e.g MUSR - 62260)
3. Test that everything behaves normally e.g, change plot type, Rebin
4. Test that the Grouping tab updates the plot
5. Tile the plots by group 
6. Load in multiple runs e.g 62260-5
7. Tile by run
8. Test that the functionality, e.g fitting, rebinning behaves normally.
```
More generally, just mess around with the GUI, there's a lot of functionality present which needs testing.
 

Fixes #26884 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
I have had a quick play with this. I think we should set it up to auto select pairs if nothing is ticked, it no pairs are available to then select all groups.

**Comment by AnthonyLim23** [comment]
Testing:

1. Cannot edit the alpha or detector ID's in the grouping tab
2. If I change the pair or group (e.g. press guess alpha) the plot does not update
3. If I plot a pair and change the plot type to counts I get an error in the logger
```Traceback (most recent call last):
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\plotting_widget\plotting_widget_presenter.py", line 133, in handle_plot_type_changed
    self._view.plot_selector.setCurrentText(ASYMMETRY_PLOT_TYPE)
AttributeError: 'QComboBox' object has no attribute 'setCurrentText'
```
4. We might want to plot the phase quad output
5. In FDA I cannot get the plot to show any of the frequency plots.

**Comment by AnthonyLim23** [comment]
I have updated the above list to cross off the completed things. Another one is:

6. To allow pairs and groups asymmetry to be plotted at the same time

**Comment by StephenSmith25** [comment]
Yes, this is ready for review now.

**Review by cailafinn — COMMENTED** [review]
Really nice to use and code looks good. Just a few minor code comments.

File: scripts/test/Muon/grouping_tab/grouping_tab_presenter_test.py [hunk 1]
```diff
@@ -17,26 +21,44 @@
...
 
-        self.context = setup_context(False)
+        setup_context_for_tests(self)
 
-        self.data_context =self.context.data_context 
-        self.gui_context = self.context.gui_context
-        self.group_context=self.context.group_pair_context  
+        # self.data_context = self.context.data_context
+        # self.gui_context = self.context.gui_context
+        # self.group_context = self.context.group_pair_context
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply scripts/test/Muon/grouping_tab/grouping_tab_presenter_test.py]: Good spot, no reason to leave them in, so they are now deleted

File: scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py [hunk 2]
```diff
@@ -21,6 +22,13 @@ class GroupingTabPresenter(object):
     functionality which covers both groups/pairs ; e.g. loading/saving/updating data.
     """
 
+    @staticmethod
+    def string_to_list(text):
+        # if text == "":
+        #     return []
+        # return [int(i) for i in text.split(",")]
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py]: Nope, appears to be the remnants of an implementation - its now removed.

File: scripts/Muon/GUI/Common/contexts/muon_context_ADS_observer.py [hunk 3]
```diff
@@ -43,11 +43,11 @@ def __init__(self, delete_callback, clear_callback, replace_callback):
     def deleteHandle(self, workspace_name, workspace):
         """
         Called when the ADS has deleted a workspace. Removes that workspace from the context and cleans up.
-        :param workspace_name: The name of the workspace
-        :param workspace: not used
+        :param workspace_name: The name of the workspace (not used)
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/contexts/muon_context_ADS_observer.py]: I think the function has to be of the form (self,workspace_name,workspace) so as to match the interface, so I don't think I can delete the unused parameters.

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py [hunk 4]
```diff
@@ -149,31 +152,35 @@ def notify_data_changed(self):
...
+                    item.setCheckState(QtCore.Qt.Checked)
+                else:
+                    item.setCheckState(QtCore.Qt.Unchecked)
+            if group_table_columns[i] == 'detector_ids':
+                # column 2 : detector IDs
                 detector_widget = table_utils.ValidatedTableItem()
                 detector_widget.setText(entry)
-                self.grouping_table.setItem(row_position, 1, detector_widget)
-                self.grouping_table.item(row_position, 1).setToolTip(entry)
+                self.grouping_table.setItem(row_position, 2, detector_widget)
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: Could we define these column numbers as variables at the top of the file to reduce chances of error later and make the code clearer.

File: scripts/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_view.py [hunk 5]
```diff
@@ -361,3 +383,15 @@ def _enable_all_buttons(self):
     def _disable_all_buttons(self):
         self.add_pair_button.setEnabled(False)
         self.remove_pair_button.setEnabled(False)
+
+    def set_to_analyse_state(self, row, state):
+        checked_state = QtCore.Qt.Checked if state is True else QtCore.Qt.Unchecked
+        item = self.get_table_item(row, 1)
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_view.py]: As before could we define which column is which at the top.

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py [hunk 6]
```diff
@@ -7,202 +7,238 @@
...
-        This property is needed to check whether all the workspaces contained in the list are still on the graph. They can
-        be removed from the graph from the figure window without this class knowing.
-        :return:
+        A list of the plotted workspaces in the figure
         """
-        self._plotted_workspaces = [item for item in self._plotted_workspaces if item in
-                                    self.plot_figure.gca().tracked_workspaces.keys()]
         return self._plotted_workspaces
 
+    # NOTE: Not sure if this property is necessary, or used for anything
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py]: Could we check and remove it if it is not required?

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py [hunk 7]
```diff
@@ -7,202 +7,238 @@
...
+            if new_top_i > new_top:
+                new_top = new_top_i
+
+        return new_bottom, new_top
+
+    def _remove_all_data_workspaces_from_plot(self, axes):
         workspaces_to_remove = self.plotted_workspaces
-        for workspace in workspaces_to_remove:
-            self.remove_workpace_from_plot(workspace)
+        for i, workspace in enumerate(workspaces_to_remove):
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py]: Why do we need to enumerate here. As far as I can tell the i isn't used for anything?

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py [hunk 8]
```diff
@@ -47,39 +52,50 @@ def setup_interface(self):
...
-        self.plot_button.setSizePolicy(QtWidgets.QSizePolicy.Fixed,
-                                       QtWidgets.QSizePolicy.Fixed)
+
+        self.plot_type_selector = QtWidgets.QCheckBox(parent=self)
+        self.plot_type_selector.setChecked(False)
+        self.plot_type_selector.setObjectName("plotTypeSelector")
+        self.plot_type_selector.setSizePolicy(QtWidgets.QSizePolicy.Fixed,
+                                              QtWidgets.QSizePolicy.Fixed)
+
+        # self.plot_button = QtWidgets.QPushButton(self)
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py]: Can we remove commented out code if it is not needed.

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py [hunk 9]
```diff
@@ -47,39 +52,50 @@ def setup_interface(self):
...
-        self.keep.hide()
+        self.horizontal_layout.addWidget(self.plot_type_selector)
+        self.horizontal_layout.addWidget(self.tiled_type_label)
+        self.horizontal_layout.addWidget(self.tile_type_selector)
+        self.horizontal_layout.addSpacing(15)
         self.horizontal_layout.addStretch(0)
         self.horizontal_layout.addWidget(self.raw)
         self.horizontal_layout.addStretch(0)
-        self.horizontal_layout.addWidget(self.plot_button)
+        # self.horizontal_layout.addWidget(self.plot_button)
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py]: Remove if not needed

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py [hunk 10]
```diff
@@ -138,19 +165,65 @@ def get_selected(self):
         return self.plot_selector.currentText()
 
     def on_plot_button_clicked(self, slot):
-        self.plot_button.clicked.connect(slot)
+        pass
```
Comment (Mantid-Matthew) [inline scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py]: If the plot button has been removed do we need this function at all anymore?

**Review by Matthew-Andrew — COMMENTED** [review]
In general code changes look good. Great job. Just a few minor comments below.

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py (hunk 7 above)
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py]: Good spot, I used to use the variable, but it is now no longer needed.

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py (hunk 10 above)
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/plotting_widget/plotting_widget_view.py]: I was unsure at the time as to whether we needed it, I've now fully removed it.

File: scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py (hunk 6 above)
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/plotting_widget/plotting_widget_model.py]: The property is used somewhere, so I'm going to decide that this is still needed, and therefore i have removed the comment.

**Review by cailafinn — CHANGES_REQUESTED** [review]
I'm happy with the code and the general workings, but there were a couple of things I found when testing again:
1. If one of the default groups is removed from the list using the `-` button, a blank plot stays on the plotting window, taking up space.
2. This is probably unrelated to this PR but is a general usability thing, the `-` button in the grouping tab only removes the bottom group, to me it should remove the highlighted group. However, that may not be required by your actual users.

**Comment by StephenSmith25** [comment]
> I'm happy with the code and the general workings, but there were a couple of things I found when testing again:
> 
>     1. If one of the default groups is removed from the list using the `-` button, a blank plot stays on the plotting window, taking up space.
> 
>     2. This is probably unrelated to this PR but is a general usability thing, the `-` button in the grouping tab only removes the bottom group, to me it should remove the highlighted group. However, that may not be required by your actual users.

Good spot, I had overlooked this completely, it should now work as intended.

**Review by cailafinn — APPROVED** [review]
I'm happy with this now. Code looks good and functions well. Great work :+1:

**Pull request merged**